walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,125 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Save all attendees that should attend and the duration of the meeting.
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long duration = request.getDuration();
+
+    // Add all time ranges that someone who is in the meeting will be in to the correct
+    // mandatory, and optional and mandatory time range arrayLists.
+    ArrayList<TimeRange> mandatoryTimeRanges = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> optionalMandatoryTimeRanges = new ArrayList<TimeRange>();
+    Iterator<Event> eventsIterator = events.iterator();
+    while (eventsIterator.hasNext()) {
+      Event event = eventsIterator.next();
+      boolean addedToOptional = false;
+      for (String eventAttendee: event.getAttendees()) {
+        if (attendees.contains(eventAttendee)) {
+          mandatoryTimeRanges.add(event.getWhen());
+          optionalMandatoryTimeRanges.add(event.getWhen());
+          break;  
+        }
+        if (!addedToOptional && optionalAttendees.contains(eventAttendee)) {
+          optionalMandatoryTimeRanges.add(event.getWhen());
+          addedToOptional = true;
+        }
```

Comment:
This logic is a bit confusing to me, but it's probably because I'm not super familiar with the goal of the project, or this class :)

Will you add more comments? Especially around the logic on line 47? I don't get what's going on with addedToOptional immediately.

Reply:
If mandatoryTimeRanges represent the times that are already blocked off for the mandatory attendees, could you rename to "mandatoryBusyTimeRanges"? This may help clear it up?

Also "optionalMandatoryTimeRanges" is a bit of an oxymoron lol. Can you just store the optional attendees busy time ranges here? Why the need to store mandatory in both arrarys?

Reply:
My bad, I'll fix that. The optionalMandatory name was meant to mean optionalAndMandatory, which i switched it to.

Reply:
I also explained that one part better and explained some other comments.